test/test.js

```diff
@@ -5,111 +5,598 @@ suite('GaiaHeader', function() {
+        this.fontFit.reset();
+
+        // Removing the action should
+        // hide the action-button, so
+        // font-fit needs to be run.
+        el.action = null;
+
+        return afterNext(el, 'runFontFit').then(() => {
+          sinon.assert.calledOnce(this.fontFit);
+        });
+      }).then(done, done);
+    });
+
+    test('It runs font-fit when a button is hidden', function(done) {
+      this.dom.innerHTML = `<gaia-header action="back">
+        <h1>foo</h1>
+        <button></button>
+      </gaia-header>`;
+
+      var el = this.dom.firstElementChild;
+      var button = el.querySelector('button');
+
+      afterNext(el, 'runFontFit').then(() => {
+        this.fontFit.reset();
+
+        // Hiding a button triggers
+        // the mutation observer
+        button.hidden = true;
+
+        return afterNext(el, 'runFontFit').then(() => {
+          sinon.assert.calledOnce(this.fontFit);
+        });
+      }).then(done, done);
+    });
+
+    test('It runs when a button\'s classList changes', function(done) {
+      this.dom.innerHTML = `<gaia-header action="back">
+        <h1>foo</h1>
+        <button></button>
+      </gaia-header>`;
+
+      var el = this.dom.firstElementChild;
+      var button = el.querySelector('button');
+
+      afterNext(el, 'runFontFit').then(() => {
+        this.runFontFit.reset();
+
+        // Hiding a button triggers
+        // the mutation observer
+        button.classList.add('hidden');
+
+        return afterNext(el, 'runFontFit').then(() => {
+          sinon.assert.calledOnce(this.runFontFit);
+        });
+      }).then(done, done);
+    });
+  });
+
+  suite('title-start and title-end attributes', function() {
+    var h1, el;
+
+    function insertHeader(container, attrs, done) {
+      done = done || function() {};
+      attrs = attrs || {};
+
+      var start = 'titleStart' in attrs ? 'title-start="' +  attrs.titleStart + '"' : '';
+      var end = 'titleEnd' in attrs ? 'title-end="' +  attrs.titleEnd + '"': '';
+
+      container.innerHTML = `
+        <gaia-header action="back" ${start} ${end}>
+          <h1>Title</h1>
+        </gaia-header>
+      `;
+
+      el = container.firstElementChild;
+      h1 = el.querySelector('h1');
+
+      return afterNext(el, 'runFontFit').then(done);
+    }
+
+    suite('normal cases', function() {
+      setup(function(done) {
+        insertHeader(this.dom, { titleStart: 50, titleEnd: 100 }, done);
+      });
+
+      test('are correctly taken into account', function() {
+        assert.equal(h1.style.MozMarginStart, '50px');
+      });
+
+      test('changing start attribute is taken into account', function(done) {
+
+        afterNext(el, 'runFontFit').then(() => {
+          afterNext(el, 'runFontFit').then(() => {
+            assert.equal(h1.style.MozMarginStart, '100px');
+          }).then(done, done);
+
+          assert.equal(h1.style.MozMarginStart, '100px');
+          el.removeAttribute('title-start');
+        });
```

Comment:
I think you can unindent this somewhat, returning afterNext's return value instead of adding a `then` handler inside the other `then` handler.